File: 10_flutter_and_native_samples/android/app/src/main/kotlin/com/example/flutter_and_native_samples/CallNativeViewChannel.kt
package com.example.flutter_and_native_samples

import android.app.Activity
import android.content.Intent
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * MethodChannel bridge that starts [ProfileActivity].
 *
 * Flutter requests navigation; Android owns the full-screen native UI.
 */
class CallNativeViewChannel(
    private val activity: Activity,
    messenger: BinaryMessenger,
) : MethodChannel.MethodCallHandler {
    private val channel =
        MethodChannel(messenger, "examples.flutter_native_calls/call_native_view")

    fun register() {
        channel.setMethodCallHandler(this)
    }

    fun unregister() {
        channel.setMethodCallHandler(null)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "openProfile" -> openProfile(result)
            else -> result.notImplemented()
        }
    }

    private fun openProfile(result: MethodChannel.Result) {
        try {
            activity.startActivity(Intent(activity, ProfileActivity::class.java))
            result.success(null)
        } catch (error: Exception) {
            result.error(
                "unavailable",
                "Unable to open native profile Activity: ${error.message}",
                null,
            )
        }
    }
}

File: 10_flutter_and_native_samples/ios/Runner/CallNativeViewChannel.swift
import Flutter
import UIKit

/// MethodChannel bridge that presents [ProfileViewController].
///
/// Flutter requests navigation; iOS owns the full-screen native UI.
final class CallNativeViewChannel {
  static let channelName = "examples.flutter_native_calls/call_native_view"

  private let channel: FlutterMethodChannel

  init(messenger: FlutterBinaryMessenger) {
    channel = FlutterMethodChannel(name: Self.channelName, binaryMessenger: messenger)
  }

  func register() {
    channel.setMethodCallHandler { [weak self] call, result in
      self?.handle(call: call, result: result)
    }
  }

  func unregister() {
    channel.setMethodCallHandler(nil)
  }

  private func handle(call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "openProfile":
      openProfile(result: result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func openProfile(result: @escaping FlutterResult) {
    DispatchQueue.main.async {
      guard let presenter = Self.topViewController() else {
        result(
          FlutterError(
            code: "unavailable",
            message: "Unable to find a view controller to present the native profile.",
            details: nil
          )
        )
        return
      }

      let profile = ProfileViewController()
      profile.modalPresentationStyle = .pageSheet
      presenter.present(profile, animated: true) {
        result(nil)
      }
    }
  }

  private static func topViewController(
    base: UIViewController? = {
      let scenes = UIApplication.shared.connectedScenes.compactMap { $0 as? UIWindowScene }
      let window = scenes
        .flatMap(\.windows)
        .first(where: \.isKeyWindow)
      return window?.rootViewController
    }()
  ) -> UIViewController? {
    if let navigation = base as? UINavigationController {
      return topViewController(base: navigation.visibleViewController)
    }
    if let tab = base as? UITabBarController {
      return topViewController(base: tab.selectedViewController)
    }
    if let presented = base?.presentedViewController {
      return topViewController(base: presented)
    }
    return base
  }
}
